[CI] Workaround to the dind-rootless limitation to restore user on build.sh and test.sh

jeanschmidt opened the pull request.

jeanschmidt committed:
Workarround to the dind-rootless limitation to restore user on build.sh and test.sh

zxiiro commented on .ci/pytorch/build.sh:
```diff
@@ -223,8 +223,10 @@ if [[ "${BUILD_ENVIRONMENT}" != *android* && "${BUILD_ENVIRONMENT}" != *cuda* ]]
+# Workaround for dind-rootless userid mapping (https://github.com/pytorch/ci-infra/issues/96)
 WORKSPACE_ORIGINAL_OWNER_ID=$(stat -c '%u' "/var/lib/jenkins/workspace")
 sudo chown -R jenkins /var/lib/jenkins/workspace
+trap "sudo chown -R $WORKSPACE_ORIGINAL_OWNER_ID /var/lib/jenkins/workspace" EXIT
```
We should run the trap command before `sudo chown -R jenkins /var/lib/jenkins/workspace` so that we can ensure it runs even if the chown command fails.

zxiiro replied:
I updated the PR to move the trap up one line.

zxiiro committed:
Set trap before chown -u jenkins

Ensure that the trap will run even if `chown -u jenkins` fails for
some reason.

zxiiro commented:
> Looks like the trap doesn't completely work we get a sudo error when it runs:
> 
> ```
> sudo chown -R 0 /var/lib/jenkins/workspace
> sudo: setrlimit(RLIMIT_STACK): Operation not permitted
> ```
> 
> I'll investigate and see if there's anything we can do to resolve the Operation not permitted issue.

Looks like this is just a warning message and the chown command is still running despite the message. Unfortunately the workaround to disable the message didn't work for me.

https://github.com/sudo-project/sudo/issues/42

zxiiro commented:
> Looks like this is just a warning message and the chown command is still running despite the message. Unfortunately the workaround to disable the message didn't work for me.
> 
> [sudo-project/sudo#42](https://github.com/sudo-project/sudo/issues/42)

We should print a message to state that the warning message can be ignored as others troubleshooting might think this is the cause of a failure in their builds / tests.

zxiiro commented:
Looks like you can only use 1 trap at a time and there are other declared traps. We should use the trap_add function from common_utils to add our trap rather than replace the existing traps.

zxiiro commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot merge -i
